tests/cloudpickle_test.py

```diff
@@ -2130,6 +2137,38 @@ def test_pickle_importable_typevar(self):
         from typing import AnyStr
         assert AnyStr is pickle_depickle(AnyStr, protocol=self.protocol)
 
+    @unittest.skipIf(sys.version_info < (3, 7),
+                     "Pickling generics not supported below py37")
+    def test_generic_type(self):
+        for obj in _generic_objects_to_test():
+            pickle_depickle(obj, protocol=self.protocol)
+
+    @unittest.skipIf(sys.version_info < (3, 7),
+                     "Pickling type hints not supported below py37")
+    def test_locally_defined_class_with_type_hints(self):
+        with subprocess_worker(protocol=self.protocol) as worker:
+            for type_ in _generic_objects_to_test():
```

Comment:
I would be in favor of using this test to check not only generic types, but all kinds of type annotations, including simple literals from the typing module as was done before.

Reply:
Trying to understand here: why would you want the pickling of these constructs to be tested in the same, single test(and not in `test_pickling_stlib_typing_module_globals` or something like that)? The pickling process and branches traversed to pickle `C` (which is the only purpose of this PR) is not the same as globals defined in the `typing` module. To the best of my knowledge, there is no reason for why pickling these objects should be tested in a single unit-test.

Reply:
All type annotations should be preserved, dynamic or non-dynamic.

We could have additional tests for things that are specific to dynamic ones but we should check that we do not break the pickling of annotations with non-dynamic types too, no?

Reply:
I agree. I simply want [my/your/any-cloudpickle-contributor/maintainer] future-self to be able to quickly understand why a specific line/commit exists, and this through atomic unit tests (we can also have composite ones, but to me atomic ones are **necessary**), related to the exact feature we decided to introduce in the said line/commit. In our case, this PR introduces support for pickling classes **subclassing** `typing.Generic` constructs. I think we should have an atomic unit test for this feature. 
- If we are missing tests testing the pickling of `typing.Generic` constructs only, let's add another test.
- If we are missing tests testing the pickling of the `globals` of the `typing` module, let's add another test.
- If we are missing composite tests testing the pickling what's inside a dynamic class/function `__annotations__`  (be it `Generic` construct or other `typing` constructs), let's add another test.
- If we want to test any combination of the 4 situations above (the first one being subclassing `typing.Generic`), then let's add another test.

 But why mixing everything up in one single test? The test suite of `cloudpickle` stills run pretty fast (<1 minute against `CPython`), not sure there is a need for optimization on this end yet.

Two final remarks:
- I hope I don't  sound like I'm bikeshedding -- I just want to make sure every one is on the same line when hitting `merge`.
- If this message, along with the other ones I wrote about this topic, did not convince you, then I'll withdraw. I can also write a PR by myself once we merge this PR and #318 if we don't want to spend too much time reviewing these PRs.

Reply:
+1 for an atomic test for public Generic type subclasses alone (without the subprocess thing). But I like to also have a big integration test that test dynamic classes annotated with all the possible kinds of types (linear, typevars, generic and generic subclasses) as a big integration test for this.